crates/stages/src/stages/headers.rs

```diff
@@ -218,6 +221,61 @@ where
 
         info!(target: "sync::stages::headers", len = downloaded_headers.len(), "Received headers");
 
+        let tip_block_number = match tip {
+            // If tip is hash and it equals to the first downloaded header's hash, we can use
+            // the block number of this header as tip.
+            BlockHashOrNumber::Hash(hash) => downloaded_headers.first().and_then(|header| {
+                if header.hash == hash {
+                    Some(header.number)
+                } else {
+                    None
+                }
+            }),
+            // If tip is number, we can just grab it and not resolve using downloaded headers.
+            BlockHashOrNumber::Number(number) => Some(number),
+        };
```

Comment:
I can see how `.map_(number|hash)() ->(num|hash)` functions could be useful.

feel free to open a good first issue for these combinators, perhaps some will pick that up